README.rst
******************************
Libravatar MediaWiki extension
******************************

Provides a ``<libravatar email="foo@example.org"/>`` tag for MediaWiki__
which displays the avatar image for a given email address,
using the Libravatar__ service.
It is federated and falls back to Gravatar if no Libravatar is found.

See the Homepage__ and `source code`__.

__ http://www.mediawiki.org/
__ https://www.libravatar.org/
__ http://www.mediawiki.org/wiki/Extension:Libravatar
__ https://github.com/phispi/mediawiki-libravatar


Examples
========
Simple avatar::

  <libravatar email="foo@example.org"/>

Avatar with special size::

  <libravatar email="foo@example.org" size="32"/>

Image title::

  <libravatar email="foo@example.org" title="Foo's avatar"/>

Verbose notation and default avatar::

  <libravatar default="monsterid">foo@example.org</libravtar>

MediaWiki user's email address::

  <libravatar user="foo"/>


Supported attributes
====================
``email``
  Email address to show avatar of. Required (if no user is given)

``user``
  MediaWiki user name of an existing MediaWiki user having an attached email address.

``size``
  Size in pixels (avatars are quadratic)

``default``
  Default icon type when no avatar has been found.
  Options: 404, mm, identicon, monsterid, wavatar, retro

``algorithm``
  Email hashing algorithm. md5 or sha256

``alt``
  "alt" attribute value for the image tag

``title``
  "title" attribute value for the image tag

``style``
  "style" attribute value for the image tag

``class``
  "class" attribute value for the image tag


Global configuration variables
==============================
- ``$wgLibravatarSize``
- ``$wgLibravatarDefault``
- ``$wgLibravatarAlgorithm``



Installation
============
Phar
----
Copy the ``mediawiki-libravatar-0.2.0.phar`` file into your MediaWiki
``extensions/`` directory and add the following to your
``LocalSettings.php``::

  require_once "$IP/extensions/mediawiki-libravatar-0.2.0.phar";

That's it, it works now. All dependencies are included in the ``.phar`` file.

If you get a white page, then you're `probably running Suhosin`__.

__ http://cweiske.de/tagebuch/suhosin-phar.htm


Normal installation
-------------------

1. Run $ pear install Services_Libravatar-alpha
2. Copy this extension's ``Libravatar.php``, ``Libravatar.body.php``
   and ``Libravatar.i18n.php`` into the directory
   ``$mediawiki/extensions/Libravatar/``
3. Edit ``LocalSettings.php``::

    require_once "$IP/extensions/Libravatar/Libravatar.php";


Dependencies
============
Requires the `Services_Libravatar`__ package from PEAR.

__ http://pear.php.net/package/Services_Libravatar


Author
======

Christian Weiske, christian.weiske@netresearch.de
Philipp Spitzer, philipp.spitzer@winterrodeln.org


ChangeLog
2012-10-04  Philipp Spitzer  <philipp.spitzer@winterrodeln.org>

	* Version 0.2.0
	* MediaWiki user can be given instead of an email address.
	* "alt" attribute can be changed now.
	* Support of "class" attribute
	* Validation of email addresses added.
	* Internationalization: Currently English and German are supported.
	* Minor changes to apply to MediaWiki coding standards:
	Now we have three .php files instead of one.
	* Fixed bug: The attribute "width" was generated twice.

2012-07-31  Christian Weiske  <christian.weiske@netresearch.de>

	* First version 0.1.0



Libravatar.body.php
<?php

class LibravatarExtension {
    /**
     * Generates the output for a <libravatar/> tag during parsing.
     *
     * @param string      $content Content of the tag
     *                             (between opening and closing tag)
     * @param array       $params  Array of tag parameters.
     * @param Parser      $parser  MediaWiki parser object
     * @param PPFrame_DOM $frame   Context data with e.g. template variables
     *
     * @return string HTML representation of the libravatar tag.
     */
    function render($content, $params, $parser, $frame) {
        // setup variables
        global $wgLibravatarSize;
        global $wgLibravatarDefault;
        global $wgLibravatarAlgorithm;


        // parse attributes
        try {
            // user attribute (optional)
            $user = null;
            if (isset($params['user']))
                $user = $parser->recursiveTagParse($params['user'], $frame);

            // email attribute (mandatory if no user attribute is given)
            $email = null;
            if (isset($params['email']))  $email = $parser->recursiveTagParse($params['email'], $frame);
            elseif (trim($content) != '') $email = $parser->recursiveTagParse(trim($content), $frame);
            elseif (!is_null($user)) {
                // take email from MediaWiki user
                $mwuser = User::newFromName($user);
                // if the MediaWiki user is invalid or does not exist we throw an exception
                if ($mwuser === false) throw new InvalidArgumentException(wfMessage('libravatar-invalidusername', $user)->text());
                if ($mwuser->getId() == 0) throw new InvalidArgumentException(wfMessage('libravatar-userunknown')->text());
                $email = $mwuser->getEmail();
            } else throw new InvalidArgumentException(wfMessage('libravatar-noemail')->text());

            // validate email address
            if (!Sanitizer::validateEmail($email)) throw new InvalidArgumentException(wfMessage('libravatar-invalidemail')->text());

            // size attribute (optional)
            $size = (int) $wgLibravatarSize; // default size
            if (isset($params['size'])) $size = (int) $parser->recursiveTagParse($params['size'], $frame);

            // default attribute (optional)
            $default = $wgLibravatarDefault;
            if (isset($params['default'])) $default = $parser->recursiveTagParse($params['default'], $frame);

            // algorithm attribute (optional)
            $algorithm = $wgLibravatarAlgorithm;
            if (isset($params['algorithm'])) $algorithm = $parser->recursiveTagParse($params['algorithm'], $frame);
            
            // alt attribute (optional)
            $alt = null;
            if (isset($params['alt'])) {
                $alt = $parser->recursiveTagParse($params['alt'], $frame);
            } elseif (is_null($user)) {
                $alt = wfMessage('libravatar-avatarof', str_replace(array('@', '.'), array(' at ', ' dot '), $email))->text();
            } else {
                $alt = wfMessage('libravatar-avatarof', $user)->text();
            }

            // title attribute (optional)
            $title = null;
            if (isset($params['title'])) {
                $title = $parser->recursiveTagParse($params['title'], $frame);
            }

            // class attribute (optional)
            $class = null;
            if (isset($params['class'])) {
                $class = $parser->recursiveTagParse($params['class'], $frame);
            }

            // style attribute (optional)
            $style = null;
            if (isset($params['style'])) {
                $style = $parser->recursiveTagParse($params['style'], $frame);
            }

        } catch (Exception $e) {
            return sprintf(
                '<span class="error">%s</span>',
                wfMessage('libravatar-error', $e->getMessage())->escaped()
            );
        }

        
        // use Services_Libravatar library to get avatar URL
        $sla = new Services_Libravatar();
        $sla->detectHttps();
        $sla->setSize($size);
        $sla->setDefault($default);
        $sla->setAlgorithm($algorithm);
        $url = $sla->getUrl($email);


        // convert to HTML <img ... /> tag
        $doc = new DOMDocument();
        $img = $doc->appendChild($doc->createElement('img'));
        $img->setAttribute('src', $url);
        $img->setAttribute('alt', $alt);
        $img->setAttribute('width', sprintf('%d', $size));
        $img->setAttribute('height', sprintf('%d', $size));
        if (!is_null($title)) $img->setAttribute('title', $title);
        if (!is_null($class)) $img->setAttribute('class', $class);
        if (!is_null($style)) $img->setAttribute('style', $style);
        $html = $doc->saveHTML($img);


        // return result (markerType => nowiki prevents wiki formatting of the result)
        return array($html, 'markerType' => 'nowiki');
    }

}

?>


Libravatar.i18n.php
<?php
// Internationalisation file for extension Libravatar.
$messages = array();


// There are different types of entries. All are expected to be UTF-8 encoded.
// * Text: In principle, all characters are allowed and will be escaped before showing them.
// * HTML: This entries are _not_ escaped therefore HTML tags can be used but special characters as '<' have to be escaped here with htmlspecialchars.
// * WikiText: WikiText syntax is allowed and interpreted by MediaWiki.

$messages['qqq'] = array(
    'libravatar-description' => 'Description of the extension. WikiText.',
    'libravatar-error' => 'Text to wrap a Libravatar error message. Text.',
    'libravatar-invalidusername' => 'Error message that appears when a username with invalid characters is given. Text.',
    'libravatar-userunknown' => 'Error message that appears when a user name of a non-existing mediawiki user is given. Text.',
    'libravatar-noemail' => 'Error message that appears when no email address (and no user) is given. Text.',
    'libravatar-invalidemail' => 'Error message that appears when the given email address is no validly formatted email address. Text.',
    'libravatar-avatarof' => 'Text appearing in the alt tag of the image saying that this is the avatar of $1. Text.'
);


$messages['en'] = array(
    'libravatar-description' => 'Embed gravatars from libravatar.org',
    'libravatar-error' => 'Libravatar error: $1',
    'libravatar-invalidusername' => 'username "$1" has invalid characters',
    'libravatar-userunknown' => 'user does not exist',
    'libravatar-noemail' => 'email (or user) attribute missing',
    'libravatar-invalidemail' => 'email address invalid.',
    'libravatar-avatarof' => 'Avatar of $1'
);


$messages['de'] = array(
    'libravatar-description' => 'Gravatare von libravatar.org einbinden.',
    'libravatar-error' => 'Libravatar Fehler: $1',
    'libravatar-invalidusername' => 'Der Benutzername "$1" enthält ungültige Zeichen.',
    'libravatar-userunknown' => 'Der Benutzer existiert nicht.',
    'libravatar-noemail' => 'Es wurde keine E-Mailadresse (bzw. kein Benutzername) angegeben.',
    'libravatar-invalidemail' => 'Die angegebene E-Mailadresse ist ungültig.',
    'libravatar-avatarof' => 'Avatar von $1'
);


$messages['de-formal'] = array(
);


?>


Libravatar.php
<?php
/**
 * Libravatar MediaWiki extension.
 *
 * PHP version 5.
 *
 * Provides a <libravatar email="foo@example.org"/> tag for MediaWiki
 * which displays the avatar image for a given email address.
 *
 * @category MediaWiki
 * @package  Libravatar
 * @author   Christian Weiske <christian.weiske@netresearch.de>
 * @author   Philipp Spitzer <philipp.spitzer@winterrodeln.org>
 * @license  AGPL http://www.gnu.org/licenses/agpl.txt
 * @link     http://www.mediawiki.org/wiki/Extension:Libravatar
 * @link     http://pear.php.net/package/Services_Libravatar/
 * @link     https://www.libravatar.org/
 */
if (!defined('MEDIAWIKI')) {
    die("This file is part of MediaWiki and is not a valid entry point\n");
}


$wgExtensionCredits['parserhook'][]= array(
    'name'           => 'Libravatar',
    'version'        => '0.2.0',
    'author'         => array('Christian Weiske', 'Philipp Spitzer'),
    'url'            => 'http://www.mediawiki.org/wiki/Extension:Libravatar',
    'descriptionmsg' => 'libravatar-description'
);


// Global variables (to be changed in LocalSettings.ini)
$wgLibravatarSize = 32; // avatar image width and height
$wgLibravatarDefault = null; // default image: '404', 'mm', 'identicon', 'monsterid', 'wavatar', 'retro' or null
$wgLibravatarAlgorithm = 'md5'; // 'md5' or 'sha256'


// Auto load class and register hook
$wgAutoloadClasses['LibravatarExtension'] = dirname(__FILE__) . '/Libravatar.body.php';
$wgAutoloadClasses['Services_Libravatar'] = dirname(__FILE__) . '/Services/Libravatar.php';
$wgHooks['ParserFirstCallInit'][] = 'libravatarParserFirstCallInit';
$wgExtensionMessagesFiles['Libravatar'] = dirname( __FILE__ ) . '/Libravatar.i18n.php';

// Register the <libravatar/> tag with the MediaWiki parser
function libravatarParserFirstCallInit($wgParser) {
    $wgParser->setHook('libravatar', 'LibravatarExtension::render');
    return true;
}

?>
